File: go/client/session.go
package client

import (
	"context"
	"io"
	"sync"

	pb "github.com/dgarson/claude-sidecar/gen/claude_sidecar/v1"
	structpb "google.golang.org/protobuf/types/known/structpb"
)

type Session struct {
	stream    pb.ClaudeSidecar_AttachSessionClient
	events    <-chan *pb.ServerEvent
	sendMu    sync.Mutex
	handlers  Handlers
	ctx       context.Context
	cancel    context.CancelFunc
	sessionID string
	mux       *eventMux
}

func newSession(
	ctx context.Context,
	sessionID string,
	stream pb.ClaudeSidecar_AttachSessionClient,
	handlers Handlers,
) *Session {
	sessionCtx, cancel := context.WithCancel(ctx)
	mux := newEventMux()
	events := mux.SubscribeAll(256).Chan()
	return &Session{
		stream:    stream,
		events:    events,
		handlers:  handlers,
		ctx:       sessionCtx,
		cancel:    cancel,
		sessionID: sessionID,
		mux:       mux,
	}
}

func (s *Session) Events() <-chan *pb.ServerEvent {
	return s.events
}

// Turns consumes the Events channel; use one or the other for a session.
func (s *Session) Turns(ctx context.Context) <-chan *Turn {
	return CollectTurns(ctx, s.events)
}

func (s *Session) Close() {
	s.cancel()
	_ = s.stream.CloseSend()
	s.mux.Close()
}

func (s *Session) Query(ctx context.Context, prompt string) (string, error) {
	requestID := newID("req")
	err := s.send(&pb.ClientEvent{
		RequestId:        requestID,
		SidecarSessionId: s.sessionID,
		Payload: &pb.ClientEvent_Query{
			Query: &pb.QueryRequest{Prompt: &pb.QueryRequest_PromptText{PromptText: prompt}},
		},
	})
	return requestID, err
}

func (s *Session) QueryTurn(ctx context.Context, prompt string) (*Turn, error) {
	result, err := s.Run(ctx, prompt)
	if err != nil {
		return nil, err
	}
	if result.Turn == nil {
		return nil, io.EOF
	}
	return result.Turn, nil
}

func (s *Session) StartInputStream(ctx context.Context) (string, string, error) {
	streamID := newID("input")
	requestID := newID("req")
	err := s.send(&pb.ClientEvent{
		RequestId:        requestID,
		SidecarSessionId: s.sessionID,
		Payload: &pb.ClientEvent_Query{
			Query: &pb.QueryRequest{Prompt: &pb.QueryRequest_InputStreamId{InputStreamId: streamID}},
		},
	})
	return requestID, streamID, err
}

func (s *Session) SendInputChunk(ctx context.Context, streamID string, event *structpb.Struct) error {
	return s.send(&pb.ClientEvent{
		SidecarSessionId: s.sessionID,
		Payload: &pb.ClientEvent_InputChunk{
			InputChunk: &pb.StreamInputChunk{
				InputStreamId: streamID,
				Event:         event,
			},
		},
	})
}

func (s *Session) SendInputEvent(ctx context.Context, streamID string, event InputEvent) error {
	payload, err := InputEventStruct(event)
	if err != nil {
		return err
	}
	return s.SendInputChunk(ctx, streamID, payload)
}

func (s *Session) SendInputMap(ctx context.Context, streamID string, event map[string]any) error {
	return s.SendInputEvent(ctx, streamID, RawInputEvent(event))
}

func (s *Session) EndInputStream(ctx context.Context, streamID string) error {
	return s.send(&pb.ClientEvent{
		SidecarSessionId: s.sessionID,
		Payload: &pb.ClientEvent_EndInput{
			EndInput: &pb.EndInputStream{InputStreamId: streamID},
		},
	})
}

func (s *Session) Interrupt(ctx context.Context) error {
	return s.send(&pb.ClientEvent{
		RequestId:        newID("req"),
		SidecarSessionId: s.sessionID,
		Payload:          &pb.ClientEvent_Interrupt{Interrupt: &pb.InterruptRequest{}},
	})
}

func (s *Session) Cancel(ctx context.Context, reason string) error {
	return s.send(&pb.ClientEvent{
		RequestId:        newID("req"),
		SidecarSessionId: s.sessionID,
		Payload:          &pb.ClientEvent_Cancel{Cancel: &pb.CancelRequest{Reason: reason}},
	})
}

func (s *Session) SetPermissionMode(ctx context.Context, mode string) error {
	return s.send(&pb.ClientEvent{
		RequestId:        newID("req"),
		SidecarSessionId: s.sessionID,
		Payload: &pb.ClientEvent_SetPermissionMode{
			SetPermissionMode: &pb.SetPermissionModeRequest{Mode: mode},
		},
	})
}

func (s *Session) SetModel(ctx context.Context, model string) error {
	return s.send(&pb.ClientEvent{
		RequestId:        newID("req"),
		SidecarSessionId: s.sessionID,
		Payload: &pb.ClientEvent_SetModel{
			SetModel: &pb.SetModelRequest{Model: model},
		},
	})
}

func (s *Session) send(event *pb.ClientEvent) error {
	if event.SidecarSessionId == "" {
		event.SidecarSessionId = s.sessionID
	}
	s.sendMu.Lock()
	defer s.sendMu.Unlock()
	return s.stream.Send(event)
}

func (s *Session) startRecvLoop() {
	go func() {
		for {
			event, err := s.stream.Recv()
			if err != nil {
				s.mux.Close()
				return
			}
			s.handleCallback(event)
			s.mux.Enqueue(event)
		}
	}()
}

func (s *Session) handleCallback(event *pb.ServerEvent) {
	switch payload := event.Payload.(type) {
	case *pb.ServerEvent_ToolRequest:
		if s.handlers.Tool == nil {
			go s.sendToolResponse(payload.ToolRequest, ToolResultError("missing tool handler"))
			return
		}
		go func(req *pb.ToolInvocationRequest) {
			result, err := s.handlers.Tool(s.ctx, req)
			if err != nil {
				result = ToolResultError(err.Error())
			}
			s.sendToolResponse(req, result)
		}(payload.ToolRequest)
	case *pb.ServerEvent_HookRequest:
		if s.handlers.Hook == nil {
			go s.sendHookResponse(payload.HookRequest, &pb.HookOutput{Continue_: boolPtr(false), StopReason: "no hook handler"})
			return
		}
		go func(req *pb.HookInvocationRequest) {
			output, err := s.handlers.Hook(s.ctx, req)
			if err != nil {
				output = &pb.HookOutput{Continue_: boolPtr(false), StopReason: err.Error()}
			}
			s.sendHookResponse(req, output)
		}(payload.HookRequest)
	case *pb.ServerEvent_PermissionRequest:
		if s.handlers.Permission == nil {
			go s.sendPermissionResponse(payload.PermissionRequest, &pb.PermissionDecision{Behavior: "deny", Reason: "no permission handler"})
			return
		}
		go func(req *pb.PermissionDecisionRequest) {
			decision, err := s.handlers.Permission(s.ctx, req)
			if err != nil {
				decision = &pb.PermissionDecision{Behavior: "deny", Reason: err.Error()}
			}
			s.sendPermissionResponse(req, decision)
		}(payload.PermissionRequest)
	}
}

func (s *Session) sendToolResponse(req *pb.ToolInvocationRequest, result *structpb.Struct) {
	_ = s.send(&pb.ClientEvent{
		Payload: &pb.ClientEvent_ToolResponse{
			ToolResponse: &pb.ToolInvocationResponse{
				InvocationId: req.InvocationId,
				ToolResult:   result,
			},
		},
	})
}

func (s *Session) sendHookResponse(req *pb.HookInvocationRequest, output *pb.HookOutput) {
	_ = s.send(&pb.ClientEvent{
		Payload: &pb.ClientEvent_HookResponse{
			HookResponse: &pb.HookInvocationResponse{
				InvocationId: req.InvocationId,
				Output:       output,
			},
		},
	})
}

func (s *Session) sendPermissionResponse(req *pb.PermissionDecisionRequest, decision *pb.PermissionDecision) {
	_ = s.send(&pb.ClientEvent{
		Payload: &pb.ClientEvent_PermissionResponse{
			PermissionResponse: &pb.PermissionDecisionResponse{
				InvocationId: req.InvocationId,
				Decision:     decision,
			},
		},
	})
}

func DecodeStruct(value *structpb.Struct) map[string]any {
	return StructToMap(value)
}

File: java/src/main/java/com/dgarson/claude/sidecar/Session.java
package com.dgarson.claude.sidecar;

import claude_sidecar.v1.Sidecar.CancelRequest;
import claude_sidecar.v1.Sidecar.ClientEvent;
import claude_sidecar.v1.Sidecar.EndInputStream;
import claude_sidecar.v1.Sidecar.HookInvocationRequest;
import claude_sidecar.v1.Sidecar.HookInvocationResponse;
import claude_sidecar.v1.Sidecar.HookOutput;
import claude_sidecar.v1.Sidecar.InterruptRequest;
import claude_sidecar.v1.Sidecar.PermissionDecision;
import claude_sidecar.v1.Sidecar.PermissionDecisionRequest;
import claude_sidecar.v1.Sidecar.PermissionDecisionResponse;
import claude_sidecar.v1.Sidecar.QueryRequest;
import claude_sidecar.v1.Sidecar.ServerEvent;
import claude_sidecar.v1.Sidecar.SetModelRequest;
import claude_sidecar.v1.Sidecar.SetPermissionModeRequest;
import claude_sidecar.v1.Sidecar.StreamInputChunk;
import claude_sidecar.v1.Sidecar.ToolInvocationRequest;
import claude_sidecar.v1.Sidecar.ToolInvocationResponse;
import com.google.protobuf.Struct;
import io.grpc.stub.StreamObserver;

import java.security.SecureRandom;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Manages a bidirectional stream to an attached sidecar session.
 * Mirrors Go's session.go.
 *
 * <p>The session dispatches tool, hook, and permission callback requests to the provided
 * {@link Handlers}, and multiplexes all other events through an {@link EventMux}.</p>
 */
public final class Session implements AutoCloseable {

    private static final Logger LOG = Logger.getLogger(Session.class.getName());
    private static final SecureRandom RANDOM = new SecureRandom();

    private final String sessionId;
    private final StreamObserver<ClientEvent> requestObserver;
    private final Handlers handlers;
    private final EventMux mux;
    private final EventMux.Subscription globalSubscription;
    private final AtomicBoolean closed = new AtomicBoolean(false);
    private final Object sendLock = new Object();

    Session(String sessionId,
            StreamObserver<ClientEvent> requestObserver,
            Handlers handlers) {
        this.sessionId = sessionId;
        this.requestObserver = requestObserver;
        this.handlers = handlers;
        this.mux = new EventMux();
        this.globalSubscription = mux.subscribeAll(256);
    }

    // -- Public API --

    /** Returns the sidecar session ID. */
    public String getSessionId() {
        return sessionId;
    }

    /**
     * Returns a blocking queue that receives all server events for this session.
     * This is the Java equivalent of Go's {@code Events() <-chan *pb.ServerEvent}.
     */
    public BlockingQueue<ServerEvent> events() {
        return globalSubscription.queue();
    }

    /**
     * Send a query and return the request ID. The caller should subscribe to events
     * (or use {@link #run} / {@link #stream}) to receive the response.
     */
    public String query(String prompt) {
        String requestId = newId("req");
        send(ClientEvent.newBuilder()
                .setRequestId(requestId)
                .setSidecarSessionId(sessionId)
                .setQuery(QueryRequest.newBuilder()
                        .setPromptText(prompt)
                        .build())
                .build());
        return requestId;
    }

    /**
     * Send a query and block until the turn completes, returning the full result.
     */
    public RunResult run(String prompt) throws Exception {
        StreamHandle handle = stream(prompt);
        return handle.result();
    }

    /**
     * Send a query and return a {@link StreamHandle} for async consumption of events.
     */
    public StreamHandle stream(String prompt) {
        String requestId = query(prompt);
        EventMux.Subscription sub = mux.subscribeRequest(requestId, 256);
        return new StreamHandle(requestId, sub, mux);
    }

    /**
     * Start an input stream for streaming input events. Returns [requestId, streamId].
     */
    public String[] startInputStream() {
        String streamId = newId("input");
        String requestId = newId("req");
        send(ClientEvent.newBuilder()
                .setRequestId(requestId)
                .setSidecarSessionId(sessionId)
                .setQuery(QueryRequest.newBuilder()
                        .setInputStreamId(streamId)
                        .build())
                .build());
        return new String[]{requestId, streamId};
    }

    /** Send a raw input chunk on the given stream. */
    public void sendInputChunk(String streamId, Struct event) {
        send(ClientEvent.newBuilder()
                .setSidecarSessionId(sessionId)
                .setInputChunk(StreamInputChunk.newBuilder()
                        .setInputStreamId(streamId)
                        .setEvent(event)
                        .build())
                .build());
    }

    /** Signal end of input for the given stream. */
    public void endInputStream(String streamId) {
        send(ClientEvent.newBuilder()
                .setSidecarSessionId(sessionId)
                .setEndInput(EndInputStream.newBuilder()
                        .setInputStreamId(streamId)
                        .build())
                .build());
    }

    /** Interrupt the current turn. */
    public void interrupt() {
        send(ClientEvent.newBuilder()
                .setRequestId(newId("req"))
                .setSidecarSessionId(sessionId)
                .setInterrupt(InterruptRequest.getDefaultInstance())
                .build());
    }

    /** Cancel the current turn with a reason. */
    public void cancel(String reason) {
        send(ClientEvent.newBuilder()
                .setRequestId(newId("req"))
                .setSidecarSessionId(sessionId)
                .setCancel(CancelRequest.newBuilder().setReason(reason).build())
                .build());
    }

    /** Set the permission mode for this session. */
    public void setPermissionMode(String mode) {
        send(ClientEvent.newBuilder()
                .setRequestId(newId("req"))
                .setSidecarSessionId(sessionId)
                .setSetPermissionMode(SetPermissionModeRequest.newBuilder()
                        .setMode(mode)
                        .build())
                .build());
    }

    /** Set the model for this session. */
    public void setModel(String model) {
        send(ClientEvent.newBuilder()
                .setRequestId(newId("req"))
                .setSidecarSessionId(sessionId)
                .setSetModel(SetModelRequest.newBuilder()
                        .setModel(model)
                        .build())
                .build());
    }

    @Override
    public void close() {
        if (closed.compareAndSet(false, true)) {
            try {
                requestObserver.onCompleted();
            } catch (Exception e) {
                LOG.log(Level.FINE, "Error closing send stream", e);
            }
            mux.close();
        }
    }

    // -- Internal: send with mutex --

    void send(ClientEvent event) {
        synchronized (sendLock) {
            requestObserver.onNext(event);
        }
    }

    // -- Internal: receive loop (called from SidecarClient after stream setup) --

    StreamObserver<ServerEvent> createResponseObserver() {
        return new StreamObserver<>() {
            @Override
            public void onNext(ServerEvent event) {
                handleCallback(event);
                mux.enqueue(event);
            }

            @Override
            public void onError(Throwable t) {
                LOG.log(Level.WARNING, "Stream error for session " + sessionId, t);
                mux.close();
            }

            @Override
            public void onCompleted() {
                mux.close();
            }
        };
    }

    // -- Internal: callback dispatch --

    private void handleCallback(ServerEvent event) {
        if (event.hasToolRequest()) {
            handleToolRequest(event.getToolRequest());
        } else if (event.hasHookRequest()) {
            handleHookRequest(event.getHookRequest());
        } else if (event.hasPermissionRequest()) {
            handlePermissionRequest(event.getPermissionRequest());
        }
    }

    private void handleToolRequest(ToolInvocationRequest request) {
        Thread.ofVirtual().name("tool-handler-" + request.getInvocationId()).start(() -> {
            Struct result;
            if (handlers.toolHandler() == null) {
                result = ToolResults.error("missing tool handler");
            } else {
                try {
                    result = handlers.toolHandler().handle(request);
                } catch (Exception e) {
                    result = ToolResults.error(e.getMessage());
                }
            }
            send(ClientEvent.newBuilder()
                    .setSidecarSessionId(sessionId)
                    .setToolResponse(ToolInvocationResponse.newBuilder()
                            .setInvocationId(request.getInvocationId())
                            .setToolResult(result)
                            .build())
                    .build());
        });
    }

    private void handleHookRequest(HookInvocationRequest request) {
        Thread.ofVirtual().name("hook-handler-" + request.getInvocationId()).start(() -> {
            HookOutput output;
            if (handlers.hookHandler() == null) {
                output = HookOutput.newBuilder()
                        .setContinue(false)
                        .setStopReason("no hook handler")
                        .build();
            } else {
                try {
                    output = handlers.hookHandler().handle(request);
                } catch (Exception e) {
                    output = HookOutput.newBuilder()
                            .setContinue(false)
                            .setStopReason(e.getMessage())
                            .build();
                }
            }
            send(ClientEvent.newBuilder()
                    .setSidecarSessionId(sessionId)
                    .setHookResponse(HookInvocationResponse.newBuilder()
                            .setInvocationId(request.getInvocationId())
                            .setOutput(output)
                            .build())
                    .build());
        });
    }

    private void handlePermissionRequest(PermissionDecisionRequest request) {
        Thread.ofVirtual().name("perm-handler-" + request.getInvocationId()).start(() -> {
            PermissionDecision decision;
            if (handlers.permissionHandler() == null) {
                decision = PermissionDecision.newBuilder()
                        .setBehavior("deny")
                        .setReason("no permission handler")
                        .build();
            } else {
                try {
                    decision = handlers.permissionHandler().handle(request);
                } catch (Exception e) {
                    decision = PermissionDecision.newBuilder()
                            .setBehavior("deny")
                            .setReason(e.getMessage())
                            .build();
                }
            }
            send(ClientEvent.newBuilder()
                    .setSidecarSessionId(sessionId)
                    .setPermissionResponse(PermissionDecisionResponse.newBuilder()
                            .setInvocationId(request.getInvocationId())
                            .setDecision(decision)
                            .build())
                    .build());
        });
    }

    // -- ID generation --

    static String newId(String prefix) {
        byte[] buf = new byte[8];
        RANDOM.nextBytes(buf);
        StringBuilder sb = new StringBuilder(prefix.length() + 1 + 16);
        sb.append(prefix).append('_');
        for (byte b : buf) {
            sb.append(String.format("%02x", b & 0xff));
        }
        return sb.toString();
    }
}
